### DISCUSSION TITLE
Fieldcollection vs Object Class

### DISCUSSION THREAD

**Discussion by cancan101** [pimcore/pimcore#17536] [Q&A / Support]
What are the limitations of a Fieldcollection vs the full Object Class? i.e are there certain types of object fields, etc that can be used on a Class but cannot be used on a Fieldcollection?

More broadly: why does Pimcore have separate (and almost duplicated) concepts of Fieldcollection and Classes rather than just unifying them and having a field type on Classes that allows a "Field Collection" of a specific object type (rather than of a specific field collection type)?

The current approach limits re-use of a class, for example if I define a class and then want to re-use it as a field collection.

**Comment by fashxp** [comment]
There are a couple of limitations in regards to (complex) data types in field collections. Also in terms of how data is loaded and filtered, there are differences. 

> More broadly: why does Pimcore have separate (and almost duplicated) concepts of Fieldcollection and Classes rather than just unifying them and having a field type on Classes that allows a "Field Collection" of a specific object type (rather than of a specific field collection type)?

Fair aspect. This is mainly how it grew. The field collections started very small and became more powerful step by step :smile:

**Reply by cancan101** [reply]
>There are a couple of limitations in regards to (complex) data types in field collections. Also in terms of how data is loaded and filtered, there are differences.

Are these limitations / difference documented anywhere?

When thinking about doing data validation (e.g. with Enterprise) is it possible to write validation rules on field collections?

>The field collections started very small and became more powerful step by step

Are there any plans to unify the two concepts?

**Comment by fashxp** [comment]
> Are these limitations / difference documented anywhere?

The best way is to check in the field definitions editor and see what data types are allowed there.

> When thinking about doing data validation (e.g. with Enterprise) is it possible to write validation rules on field collections?

With data quality management you can use the symfony expressions to define some validation rules. Depending on the use case that might be a bit clumsy (due to the additional dimensions field collections open up), so you probably also want to consider implementing a [custom rule definition](https://pimcore.com/docs/platform/Data_Quality_Management/Customization/Custom_Rule_Definition) which does exactly what you need for your data model. 

> Are there any plans to unify the two concepts?

No, not yet - as they both have there reasons to exist.
